testsuite/cluster-test/src/health/fullnode_check.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::prometheus::Prometheus;
+use crate::util::unix_timestamp_now;
+use crate::{
+    cluster::Cluster,
+    health::{HealthCheck, HealthCheckContext, ValidatorEvent},
+};
+use anyhow::{format_err, Result};
+use once_cell::sync::Lazy;
+use regex::Regex;
+use std::collections::HashMap;
+use std::time::Duration;
+
+const APP_LIBRA_VALIDATOR: &str = "libra-validator";
+const APP_LIBRA_FULLNODE: &str = "libra-fullnode";
+static VAL_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"val-(\d+)").unwrap());
+static FULLNODE_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"fn-(\d+)").unwrap());
+
+pub struct FullNodeHealthCheck {
+    cluster: Cluster,
+    prometheus: Prometheus,
+}
+
+impl FullNodeHealthCheck {
+    pub fn new(cluster: Cluster, prometheus: Prometheus) -> FullNodeHealthCheck {
+        Self {
+            cluster,
+            prometheus,
+        }
+    }
+
+    pub fn state_sync_committed_version(
+        &self,
+        start: Duration,
+        end: Duration,
+        app: &str,
+    ) -> Result<HashMap<String, f64>> {
+        libra_retrier::retry(libra_retrier::fixed_retry_strategy(1_000, 5), || {
+            self.prometheus
+                .query_range_max(
+                    format!("libra_state_sync_committed_version{{app=\"{}\"}}", app),
+                    &start,
+                    &end,
+                    10, /* step */
+                )
+                .map_err(|e| format_err!("No sync committed version data: {}", e))
+        })
+    }
+}
+
+pub fn node_index(peer_name: String) -> String {
+    if let Some(cap) = VAL_REGEX.captures(peer_name.as_ref()) {
+        if let Some(cap) = cap.get(1) {
+            return cap.as_str().to_string();
+        }
+    }
+    if let Some(cap) = FULLNODE_REGEX.captures(peer_name.as_ref()) {
+        if let Some(cap) = cap.get(1) {
+            return cap.as_str().to_string();
+        }
+    }
+    panic!(
+        "Failed to parse peer name {} into validator_index",
+        peer_name
+    )
+}
+
+pub fn convert_index(input: HashMap<String, f64>) -> HashMap<String, f64> {
+    let mut res = HashMap::new();
+    for (peer_name, version) in input {
+        let peer_name = node_index(peer_name);
+        res.insert(peer_name, version);
+    }
+    res
+}
+
+impl HealthCheck for FullNodeHealthCheck {
+    fn on_event(&mut self, _event: &ValidatorEvent, _ctx: &mut HealthCheckContext) {}
```

Comment:
Can you actually change `HealthCheck` interface to have default implementation for `on_event`, so that it does not have to be defined here?

Reply:
yes, this is a good idea